index out of range in Storage/cluster.go during unit test

```
?   	github.com/zero-os/0-Disk/nbd/ardb/command	[no test files]
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo started" module=global caller=config_api.go:254
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:33886" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster foo started" module=global caller=config_api.go:386
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig cluster for foo stopped" module=global caller=config_api.go:393
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo stopped" module=global caller=config_api.go:261
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:40110" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:38964" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo started" module=global caller=config_api.go:254
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster foo started" module=global caller=config_api.go:386
20::{"status":422,"subject":"ardb","data":{"address":"127.0.0.1:38964","db":0,"type":"primary","vdiskID":"foo"}}
t=2017-11-02T23:24:35+0000 lvl=dbug msg="updating vdisk foo' primary server #1 state to offline" module=global caller=cluster.go:193
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig cluster for foo stopped" module=global caller=config_api.go:393
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo stopped" module=global caller=config_api.go:261
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:36996" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:32953" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo started" module=global caller=config_api.go:254
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster foo started" module=global caller=config_api.go:386
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for foo received config bytes from source" module=global caller=config_api.go:270
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for foo received config bytes from source" module=global caller=config_api.go:402
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo stopped" module=global caller=config_api.go:261
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig cluster for foo stopped" module=global caller=config_api.go:393
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:46651" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:40371" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:33104" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:38812" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo started" module=global caller=config_api.go:254
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster foo started" module=global caller=config_api.go:386
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo started" module=global caller=config_api.go:254
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster bar started" module=global caller=config_api.go:386
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 0 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 1 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 2 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 3 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 4 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 5 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 6 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 7 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:45467" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:34345" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 0 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 2 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 4 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 6 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
20::{"status":422,"subject":"ardb","data":{"address":"127.0.0.1:33104","db":0,"type":"template","vdiskID":"foo"}}
t=2017-11-02T23:24:35+0000 lvl=dbug msg="updating vdisk foo' template server #1 state to offline" module=global caller=cluster.go:570
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo stopped" module=global caller=config_api.go:261
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig cluster for bar stopped" module=global caller=config_api.go:393
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo stopped" module=global caller=config_api.go:261
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig cluster for foo stopped" module=global caller=config_api.go:393
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:36973" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:38061" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:40836" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:38130" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo started" module=global caller=config_api.go:254
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster foo started" module=global caller=config_api.go:386
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo started" module=global caller=config_api.go:254
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster bar started" module=global caller=config_api.go:386
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 0 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 1 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 2 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 3 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 4 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 5 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 6 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 7 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for foo received config bytes from source" module=global caller=config_api.go:270
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for foo received config bytes from source" module=global caller=config_api.go:402
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for foo received config bytes from source" module=global caller=config_api.go:270
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for bar received config bytes from source" module=global caller=config_api.go:402
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:39170" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:38192" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 0 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 2 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 4 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 6 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo stopped" module=global caller=config_api.go:261
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig cluster for bar stopped" module=global caller=config_api.go:393
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo stopped" module=global caller=config_api.go:261
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig cluster for foo stopped" module=global caller=config_api.go:393
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:33637" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:33940" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:45969" module=global caller=ledisdb.go:70
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo started" module=global caller=config_api.go:254
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster foo started" module=global caller=config_api.go:386
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo started" module=global caller=config_api.go:254
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster bar started" module=global caller=config_api.go:386
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 0 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 1 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 2 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 3 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 4 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 5 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 6 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="block 7 not available in primary storage, but did find it in template storage" module=global caller=nondeduped.go:137
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig cluster for foo stopped" module=global caller=config_api.go:393
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo stopped" module=global caller=config_api.go:261
t=2017-11-02T23:24:35+0000 lvl=dbug msg="WatchStorageClusterConfig cluster for bar stopped" module=global caller=config_api.go:393
t=2017-11-02T23:24:35+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk foo stopped" module=global caller=config_api.go:261
20::{"status":400,"subject":"ardb","data":{"address":"127.0.0.1:33940","db":0,"type":"primary","vdiskID":"foo"}}
panic: runtime error: index out of range
goroutine 335 [running]:
github.com/zero-os/0-Disk/nbd/ardb/storage.(*PrimaryCluster).handleServerStateUpdate(...)
	/home/travis/gopath/src/github.com/zero-os/0-Disk/nbd/ardb/storage/cluster.go:199
github.com/zero-os/0-Disk/nbd/ardb/storage.(*PrimaryCluster).updateServerState(0xc420501da0, 0x1, 0x1162c01, 0x0, 0x0)
	/home/travis/gopath/src/github.com/zero-os/0-Disk/nbd/ardb/storage/cluster.go:188 +0x3b5
github.com/zero-os/0-Disk/nbd/ardb/storage.(*PrimaryCluster).doAt(0xc420501da0, 0x1, 0xc4221232b0, 0xf, 0x0, 0x0, 0x17bed80, 0xc420425320, 0x0, 0x0, ...)
	/home/travis/gopath/src/github.com/zero-os/0-Disk/nbd/ardb/storage/cluster.go:152 +0x203
github.com/zero-os/0-Disk/nbd/ardb/storage.(*PrimaryCluster).DoFor(0xc420501da0, 0x7, 0x17bed80, 0xc420425320, 0x3, 0x3, 0xc420425320, 0xc42002d300)
	/home/travis/gopath/src/github.com/zero-os/0-Disk/nbd/ardb/storage/cluster.go:73 +0x224
github.com/zero-os/0-Disk/nbd/ardb/storage.(*nonDedupedStorage).SetBlock(0xc4200b61c0, 0x7, 0xc4200f4cb8, 0x8, 0x8, 0xc420035fc8, 0xf6aaf4)
	/home/travis/gopath/src/github.com/zero-os/0-Disk/nbd/ardb/storage/nondeduped.go:69 +0x24f
github.com/zero-os/0-Disk/nbd/ardb/storage.(*nonDedupedStorage).getPrimaryOrTemplateContent.func1(0xc4200b61c0, 0x7, 0xc4202f3240)
	/home/travis/gopath/src/github.com/zero-os/0-Disk/nbd/ardb/storage/nondeduped.go:128 +0x94
created by github.com/zero-os/0-Disk/nbd/ardb/storage.(*nonDedupedStorage).getPrimaryOrTemplateContent
	/home/travis/gopath/src/github.com/zero-os/0-Disk/nbd/ardb/storage/nondeduped.go:127 +0x805
FAIL	github.com/zero-os/0-Disk/nbd/ardb/storage	0.626s
```

@ master (e3e558fa37ce310eec0a398ed62207fb37feab23)

Note sure if related, but another test failed out of the blue (probably due to async issues):

```
--- FAIL: TestTemplateServerFails (0.03s)
	Error Trace:	cluster_test.go:294
	Error:      	Not equal: 
	            	expected: *errors.errorString(&errors.errorString{s:"server is unavailable"})
	            	actual: <nil>(<nil>)
```

--- a/nbd/ardb/storage/cluster.go
+++ b/nbd/ardb/storage/cluster.go
@@ -12,681 +12,475 @@ import (
 	"github.com/zero-os/0-Disk/nbd/ardb"
 )
 
-// NewPrimaryCluster creates a new PrimaryCluster.
-// See `PrimaryCluster` for more information.
-func NewPrimaryCluster(ctx context.Context, vdiskID string, cs config.Source) (*PrimaryCluster, error) {
-	primaryCluster := &PrimaryCluster{
-		vdiskID: vdiskID,
-		pool:    ardb.NewPool(nil),
+// NewCluster creates a new storage cluster for the given vdiskID,
+// using the given controller to control the server state and fetch that state from.
+func NewCluster(vdiskID string, controller ClusterStateController) (*Cluster, error) {
+	if vdiskID == "" {
+		return nil, errors.New("storage.Cluster requires a non-nil vdiskID")
 	}
-	err := primaryCluster.spawnConfigReloader(ctx, cs)
-	if err != nil {
-		primaryCluster.Close()
-		return nil, err
+	if isInterfaceValueNil(controller) {
+		return nil, errors.New("storage.Cluster requires a non-nil ClusterStateController")
 	}
-
-	return primaryCluster, nil
+	return &Cluster{
+		vdiskID:    vdiskID,
+		pool:       ardb.NewPool(nil),
+		controller: controller,
+	}, nil
 }
 
-// PrimaryCluster defines a vdisk's primary cluster.
-// It supports hot reloading of the configuration
-// and state handling of the individual servers of a cluster.
-type PrimaryCluster struct {
-	vdiskID string
-
-	servers     []config.StorageServerConfig
-	serverCount int64
-
-	pool   *ardb.Pool
-	cancel context.CancelFunc
-
-	mux sync.RWMutex
+// Cluster defines a cluster which applies actions on servers,
+// which are configured within a state controller.
+// This state is both updated by external events (config hot reloading),
+// as well as internal events (updating a server based on state changes).
+type Cluster struct {
+	vdiskID    string
+	pool       *ardb.Pool
+	controller ClusterStateController
 }
 
 // Do implements StorageCluster.Do
-func (pc *PrimaryCluster) Do(action ardb.StorageAction) (reply interface{}, err error) {
-	pc.mux.RLock()
-	// compute server index of first available server
-	serverIndex, err := ardb.FindFirstServerIndex(pc.serverCount, pc.serverOperational)
-	if err != nil {
-		pc.mux.RUnlock()
-		return nil, err
-	}
-	cfg := pc.servers[serverIndex]
-	pc.mux.RUnlock()
+func (cluster *Cluster) Do(action ardb.StorageAction) (reply interface{}, err error) {
+	var state ServerState
+	// keep trying to apply action, until it works out,
+	// or until no server is available any longer.
+	for {
+		state, err = cluster.controller.ServerState()
+		if err != nil {
+			// server wasn't avaialable for some illegal reason,
+			// or no server was available at all
+			return nil, err
+		}
 
-	return pc.doAt(serverIndex, cfg, action)
+		// apply action, and return its results if the action was indeed applied.
+		reply, err = cluster.applyAction(&state, action)
+		if err != errActionNotApplied {
+			return reply, err
+		}
+	}
 }
 
 // DoFor implements StorageCluster.DoFor
-func (pc *PrimaryCluster) DoFor(objectIndex int64, action ardb.StorageAction) (reply interface{}, err error) {
-	pc.mux.RLock()
-	// compute server index for the server which maps to the given object index
-	serverIndex, err := ardb.ComputeServerIndex(pc.serverCount, objectIndex, pc.serverOperational)
-	if err != nil {
-		pc.mux.RUnlock()
-		return nil, err
-	}
-	cfg := pc.servers[serverIndex]
-	pc.mux.RUnlock()
+func (cluster *Cluster) DoFor(objectIndex int64, action ardb.StorageAction) (reply interface{}, err error) {
+	var state ServerState
+	// keep trying to apply action, until it works out,
+	// or until no server is available any longer.
+	for {
+		state, err = cluster.controller.ServerStateFor(objectIndex)
+		if err != nil {
+			// server wasn't avaialable for some illegal reason,
+			// or no server was available at all
+			return nil, err
+		}
 
-	return pc.doAt(serverIndex, cfg, action)
+		// apply action, and return its results if the action was indeed applied.
+		reply, err = cluster.applyAction(&state, action)
+		if err != errActionNotApplied {
+			return reply, err
+		}
+	}
 }
 
 // DoForAll implements StorageCluster.DoForAll
-func (pc *PrimaryCluster) DoForAll(pairs []ardb.IndexActionPair) ([]interface{}, error) {
-	// a shortcut in case we have received no pairs, or just a single one
-	switch len(pairs) {
-	case 0:
-		return nil, nil // nothing to do
-	case 1:
-		reply, err := pc.DoFor(pairs[0].Index, pairs[1].Action)
-		if err != nil {
-			return nil, err
-		}
-		return []interface{}{reply}, nil
-	}
-
-	pc.mux.RLock()
-	defer pc.mux.RUnlock()
-
-	// sort all actions in terms of their mapped server index
-	servers := make(map[int64]*ardb.IndexActionMap)
-	for index, pair := range pairs {
-		// compute server index which maps to the given object index
-		serverIndex, err := ardb.ComputeServerIndex(pc.serverCount, pair.Index, pc.serverOperational)
-		if err != nil {
-			return nil, err
-		}
-		// add the pair to the relevant map
-		server, ok := servers[serverIndex]
-		if !ok {
-			server = new(ardb.IndexActionMap)
-			servers[serverIndex] = server
-		}
-		server.Add(int64(index), pair.Action)
-	}
-
-	// a shortcut if we are lucky enough to only have 1 server
-	if len(servers) == 1 {
-		for serverIndex, indexActionMap := range servers {
-			return ardb.Values(pc.doAt(serverIndex, pc.servers[serverIndex],
-				ardb.Commands(indexActionMap.Actions...)))
-		}
-	}
-
-	// create result -type and -channel to collect all replies async
-	type serverResult struct {
-		ServerIndex int64
-		Replies     []interface{}
-		Error       error
-	}
-	ch := make(chan serverResult)
-
-	// apply all actions async, and collect the results
-
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	var wg sync.WaitGroup
-	for serverIndex := range servers {
-		wg.Add(1)
-
-		// local variables to be used within the goroutine scope
-		indexActionMap := servers[serverIndex]
-		result := serverResult{ServerIndex: serverIndex}
-
-		go func() {
-			defer wg.Done()
-			result.Replies, result.Error = ardb.Values(pc.doAt(
-				result.ServerIndex, pc.servers[result.ServerIndex],
-				ardb.Commands(indexActionMap.Actions...)))
-			select {
-			case ch <- result:
-			case <-ctx.Done():
-			}
-		}()
-	}
-
-	// close channel when all inpu
-	go func() {
-		wg.Wait()
-		close(ch)
-	}()
-
-	// we'll collect all replies in order
-	replies := make([]interface{}, len(pairs))
-
-	// collect all replies
-	for result := range ch {
-		// return early if an error occured
-		if result.Error != nil {
-			return nil, errors.Wrapf(result.Error,
-				"error while applying actions in serverIndex %d", result.ServerIndex)
-		}
-
-		// get the indexActionMap for the given server,
-		// such that we can retrieve the correct reply index
-		m := servers[result.ServerIndex]
-
-		// collect all received replies in order
-		for i, reply := range result.Replies {
-			replyIndex := m.Indices[i]
-			replies[replyIndex] = reply
-		}
-	}
-
-	// return all replies from all servers in ordered form
-	return replies, nil
+func (cluster *Cluster) DoForAll(pairs []ardb.IndexActionPair) ([]interface{}, error) {
+	// [TODO] either remove `DoForAll` from the StorageCluster API,
+	//        or support it in this implementation.
+	panic("DoForAll is not supported for now")
 }
 
-// ServerIterator implements StorageCluster.ServerIterator.
-func (pc *PrimaryCluster) ServerIterator(ctx context.Context) (<-chan ardb.StorageServer, error) {
-	pc.mux.Lock()
+// ServerIterator implements StorageCluster.ServerIterator
+func (cluster *Cluster) ServerIterator(ctx context.Context) (<-chan ardb.StorageServer, error) {
 	ch := make(chan ardb.StorageServer)
 	go func() {
-		defer pc.mux.Unlock()
-		defer close(ch)
+		log.Debugf("starting server iterator for vdisk %s's storage.Cluster", cluster.vdiskID)
+		defer func() {
+			close(ch) // close channel iterator when finished
+			log.Debugf("stopping server iterator for vdisk %s's storage.Cluster", cluster.vdiskID)
+		}()
 
-		for index := int64(0); index < pc.serverCount; index++ {
-			operational, _ := pc.serverOperational(index)
-			if !operational {
-				continue
-			}
-
-			server := primaryStorageServer{
-				index:   index,
-				cluster: pc,
+		for index := int64(0); index < cluster.controller.ServerCount(); index++ {
+			server := smartServer{
+				Index:   index,
+				Cluster: cluster,
 			}
 
 			select {
+			case ch <- server:
 			case <-ctx.Done():
 				return
-			case ch <- server:
 			}
 		}
 	}()
 	return ch, nil
 }
 
-// ServerCount implements StorageCluster.ServerCount.
-func (pc *PrimaryCluster) ServerCount() int64 {
-	pc.mux.RLock()
-
-	count := pc.serverCount
-	for _, server := range pc.servers {
-		if server.State != config.StorageServerStateOnline {
-			count--
-		}
-	}
-	pc.mux.RUnlock()
-
-	return count
+// ServerCount implements StorageCluster.ServerCount
+func (cluster *Cluster) ServerCount() int64 {
+	return cluster.controller.ServerCount()
 }
 
-// execute an exuction at a given primary server
-func (pc *PrimaryCluster) doAt(serverIndex int64, cfg config.StorageServerConfig, action ardb.StorageAction) (reply interface{}, err error) {
-	// establish a connection for the given config
-	conn, err := pc.pool.Dial(cfg)
+// Close this storage cluster's open resources.
+func (cluster *Cluster) Close() error {
+	var slice errors.ErrorSlice
+	slice.Add(cluster.controller.Close())
+	slice.Add(cluster.pool.Close())
+	return slice.AsError()
+}
+
+// applyAction applies the storage action to the server
+// that can be dialer for the given action.
+func (cluster *Cluster) applyAction(state *ServerState, action ardb.StorageAction) (reply interface{}, err error) {
+	if state.Config.State != config.StorageServerStateOnline {
+		return nil, ardb.ErrServerUnavailable
+	}
+
+	// try to open connection,
+	// and apply the action to that connection if it could be dialed.
+	conn, err := cluster.pool.Dial(state.Config)
 	if err == nil {
 		defer conn.Close()
 		reply, err = action.Do(conn)
 		if err == nil || errors.Cause(err) == ardb.ErrNil {
-			return
+			return reply, err
 		}
 	}
 
-	// TODO:
-	// add self-healing...
-	// see: https://github.com/zero-os/0-Disk/issues/445
-	// and  https://github.com/zero-os/0-Disk/issues/284
-
-	// an error has occured, broadcast it to AYS
-	status := MapErrorToBroadcastStatus(err)
-	log.Broadcast(
-		status,
-		log.SubjectStorage,
-		log.ARDBServerTimeoutBody{
-			Address:  cfg.Address,
-			Database: cfg.Database,
-			Type:     log.ARDBPrimaryServer,
-			VdiskID:  pc.vdiskID,
-		},
-	)
-
-	// mark server as offline, so that next time this server will trigger an error,
-	// such that we don't broadcast all the time
-	if err := pc.updateServerState(serverIndex, config.StorageServerStateOffline); err != nil {
-		log.Errorf("couldn't update primary server (%d) state to offline: %v", serverIndex, err)
+	// mark the server as offline,
+	// as something went wrong
+	state.Config.State = config.StorageServerStateOffline
+	if cluster.controller.UpdateServerState(*state) {
+		// broadcast the error to AYS
+		status := MapErrorToBroadcastStatus(err)
+		log.Broadcast(
+			status,
+			log.SubjectStorage,
+			log.ARDBServerTimeoutBody{
+				Address:  state.Config.Address,
+				Database: state.Config.Database,
+				Type:     state.Type,
+				VdiskID:  cluster.vdiskID,
+			},
+		)
 	}
 
-	return nil, ardb.ErrServerUnavailable
+	return nil, errActionNotApplied
 }
 
-// Close any open resources
-func (pc *PrimaryCluster) Close() error {
-	pc.cancel()
-	pc.pool.Close()
-	return nil
-}
-
-// serverOperational returns true if
-// a server on the given index is available for operation.
-func (pc *PrimaryCluster) serverOperational(index int64) (bool, error) {
-	switch pc.servers[index].State {
-	case config.StorageServerStateOnline:
-		return true, nil
-
-	case config.StorageServerStateOffline:
-		return false, ardb.ErrServerUnavailable
-
-	case config.StorageServerStateRIP:
-		return false, nil
-
-	default:
-		return false, ardb.ErrServerStateNotSupported
-	}
-}
-
-func (pc *PrimaryCluster) updateServerState(index int64, state config.StorageServerState) error {
-	pc.mux.Lock()
-	defer pc.mux.Unlock()
-
-	err := pc.handleServerStateUpdate(index, state)
-	if err != nil {
-		return err
-	}
-
-	log.Debugf("updating vdisk %s' primary server #%d state to %s", pc.vdiskID, index, state)
-	pc.servers[index].State = state
-	return nil
-}
-
-func (pc *PrimaryCluster) handleServerStateUpdate(index int64, state config.StorageServerState) error {
-	if pc.servers[index].State == state {
-		return nil // nothing to do
-	}
-
-	// [TODO]
-	// Handle Online => Nothing to do ?!
-	// Handle Offline => Notify Tlog server to not sync with TLog
-	// Handle Restore => Copy data from slave to primary, Notify TLog server it can re-use that slave
-	// Handle Respread => Copy data from slave to primary, Mark server afterwards as RIP
-	// Handle RIP => nothing to do (should tlog be warned of this though?!)
-	//
-	// [TODO] should we also check the state flow? e.g. does `RIP => Online` make sense?!
-
-	switch state {
-	case config.StorageServerStateOnline, config.StorageServerStateOffline, config.StorageServerStateRIP:
-		return nil // supported
-
-	default:
-		return ardb.ErrServerStateNotSupported
-	}
-}
-
-// spawnConfigReloader starts all needed config watchers,
-// and spawns a goroutine to receive the updates.
-// An error is returned in case the initial watch-creation and config-update failed.
-// All future errors will be logged (and optionally broadcasted),
-// without stopping this goroutine.
-func (pc *PrimaryCluster) spawnConfigReloader(ctx context.Context, cs config.Source) error {
-	// create the context and cancelFunc used for the master watcher.
-	ctx, pc.cancel = context.WithCancel(ctx)
-
-	// create the master watcher if possible
-	vdiskNBDRefCh, err := config.WatchVdiskNBDConfig(ctx, cs, pc.vdiskID)
-	if err != nil {
-		return err
-	}
-	vdiskNBDConfig := <-vdiskNBDRefCh
-
-	var primaryClusterCfg config.StorageClusterConfig
-
-	// create the primary storage cluster watcher,
-	// and execute the initial config update iff
-	// an internal watcher is created.
-	var primaryWatcher ClusterConfigWatcher
-	clusterExists, err := primaryWatcher.SetClusterID(ctx, cs, pc.vdiskID, vdiskNBDConfig.StorageClusterID)
-	if err != nil {
-		return err
-	}
-	if !clusterExists {
-		panic("primary cluster should exist on a non-error path")
-	}
-	primaryClusterCfg = <-primaryWatcher.Receive()
-	err = pc.updatePrimaryStorageConfig(primaryClusterCfg)
-	if err != nil {
-		return err
-	}
-
-	// spawn the config update goroutine
-	go func() {
-		var ok bool
-		for {
-			select {
-			case <-ctx.Done():
-				return
-
-			// handle clusterID reference updates
-			case vdiskNBDConfig, ok = <-vdiskNBDRefCh:
-				if !ok {
-					return
-				}
-
-				_, err = primaryWatcher.SetClusterID(
-					ctx, cs, pc.vdiskID, vdiskNBDConfig.StorageClusterID)
-				if err != nil {
-					log.Errorf("failed to watch new primary cluster config: %v", err)
-				}
-
-			// handle primary cluster storage updates
-			case primaryClusterCfg = <-primaryWatcher.Receive():
-				err = pc.updatePrimaryStorageConfig(primaryClusterCfg)
-				if err != nil {
-					log.Errorf("failed to update new primary cluster config: %v", err)
-				}
-			}
-		}
-	}()
-
-	// all is operational, no error to return
-	return nil
-}
-
-// updatePrimaryStorageConfig overwrites
-// the currently used primary storage config,
-func (pc *PrimaryCluster) updatePrimaryStorageConfig(cfg config.StorageClusterConfig) error {
-	pc.mux.Lock()
-	defer pc.mux.Unlock()
-
-	serverCount := int64(len(cfg.Servers))
-	if serverCount > pc.serverCount {
-		serverCount = pc.serverCount
-	}
-
-	var err error
-	var origServer, newServer config.StorageServerConfig
-
-	for index := int64(0); index < serverCount; index++ {
-		origServer, newServer = pc.servers[index], cfg.Servers[index]
-		if !storageServersEqual(origServer, newServer) {
-			continue // a new server or non-changed state, so no update here
-		}
-
-		err = pc.handleServerStateUpdate(index, newServer.State)
-		if err != nil {
-			return err
-		}
-	}
-
-	pc.servers = cfg.Servers
-	pc.serverCount = int64(len(cfg.Servers))
-	return nil
-}
-
-// primaryStorageServer defines a primary storage server.
-type primaryStorageServer struct {
-	index   int64
-	cluster *PrimaryCluster
+// smartServer defines an ardb.StorageServer returned
+// by the default storage.Cluster, and applies a connection to
+// whatever server that functions first for the given server index.
+type smartServer struct {
+	Index   int64
+	Cluster *Cluster
 }
 
 // Do implements StorageServer.Do
-func (server primaryStorageServer) Do(action ardb.StorageAction) (reply interface{}, err error) {
-	cfg := server.cluster.servers[server.index]
-	return server.cluster.doAt(server.index, cfg, action)
+func (server smartServer) Do(action ardb.StorageAction) (reply interface{}, err error) {
+	var state ServerState
+	// keep trying to apply action, until it works out,
+	// or until no server is available any longer.
+	for {
+		state, err = server.Cluster.controller.ServerStateAt(server.Index)
+		if err != nil {
+			// server wasn't avaialable for some illegal reason,
+			// or no server was available at all
+			return nil, err
+		}
+
+		// apply action, and return its results if the action was indeed applied.
+		reply, err = server.Cluster.applyAction(&state, action)
+		if err != errActionNotApplied {
+			return reply, err
+		}
+	}
 }
 
 // Config implements StorageServer.Config
-func (server primaryStorageServer) Config() config.StorageServerConfig {
-	return server.cluster.servers[server.index]
+func (server smartServer) Config() config.StorageServerConfig {
+	state, err := server.Cluster.controller.ServerStateAt(server.Index)
+	if err != nil {
+		return config.StorageServerConfig{State: config.StorageServerStateRIP}
+	}
+	return state.Config
 }
 
-// NewTemplateCluster creates a new TemplateCluster.
-// See `TemplateCluster` for more information.
-func NewTemplateCluster(ctx context.Context, vdiskID string, cs config.Source) (*TemplateCluster, error) {
-	templateCluster := &TemplateCluster{
-		vdiskID: vdiskID,
-		pool:    ardb.NewPool(nil),
+// ClusterStateController is used as the internal state controller for the storage Cluster.
+// It is used to retrieve server configs and update server configs (internal and external).
+type ClusterStateController interface {
+	// Retrieve the first available server state.
+	ServerState() (state ServerState, err error)
+	// Retrieve the server state which maps to a given objectIndex.
+	ServerStateFor(objectIndex int64) (state ServerState, err error)
+	// Retrieve the server state at the given serverIndex.
+	ServerStateAt(serverIndex int64) (state ServerState, err error)
+
+	// Update the server state.
+	// An update might be ignored if it is deemed to be out of date.
+	// True is returned in case the update was applied.
+	UpdateServerState(state ServerState) bool
+
+	// ServerCount returns the (flat) amount of servers,
+	// this state has in one dimension.
+	ServerCount() int64
+
+	// Close any open resources, previously in-use by this controller.
+	Close() error
+}
+
+// ServerState is a snapshot of the state of a server,
+// as it is retrieved from a ClusterStateController.
+type ServerState struct {
+	// Index of the server within the internal cluster (model)
+	Index int64
+	// Config of the server in its current state
+	Config config.StorageServerConfig
+	// Type of the server: {primary, slave, template}
+	Type log.ARDBServerType
+}
+
+// NewPrimaryCluster creates a new PrimaryCluster.
+// This cluster type supports config hot-reloading, but no self-healing of servers.
+// This cluster type does support hot-swapping of 2 online servers (which share the same index),
+// meaning that prior to swapping, the data for this vdisk will be copied from the old to the new server.
+// See `Cluster` for more information.
+func NewPrimaryCluster(ctx context.Context, vdiskID string, cs config.Source) (*Cluster, error) {
+	controller := &singleClusterStateController{
+		vdiskID:       vdiskID,
+		optional:      false,
+		copyOnHotSwap: true,
+		configSource:  cs,
+		serverType:    log.ARDBPrimaryServer,
+		getClusterID:  getPrimaryClusterID,
 	}
-	err := templateCluster.spawnConfigReloader(ctx, cs)
+	err := controller.spawnConfigReloader(ctx, cs)
 	if err != nil {
-		templateCluster.Close()
+		controller.Close()
 		return nil, err
 	}
 
-	return templateCluster, nil
+	return NewCluster(vdiskID, controller)
 }
 
-// TemplateCluster defines a vdisk'stemplate cluster (configured or not).
-// It supports hot reloading of the configuration.
-type TemplateCluster struct {
-	vdiskID string
+// NewSlaveCluster creates a new SlaveCluster.
+// This cluster type supports config hot-reloading, but no self-healing of servers.
+// This cluster type does support hot-swapping of 2 online servers (which share the same index),
+// meaning that prior to swapping, the data for this vdisk will be copied from the old to the new server.
+// See `Cluster` for more information.
+func NewSlaveCluster(ctx context.Context, vdiskID string, optional bool, cs config.Source) (*Cluster, error) {
+	controller := &singleClusterStateController{
+		vdiskID:       vdiskID,
+		optional:      optional,
+		copyOnHotSwap: true,
+		configSource:  cs,
+		serverType:    log.ARDBSlaveServer,
+		getClusterID:  getSlaveClusterID,
+	}
+	err := controller.spawnConfigReloader(ctx, cs)
+	if err != nil {
+		controller.Close()
+		return nil, err
+	}
+
+	return NewCluster(vdiskID, controller)
+}
+
+// NewTemplateCluster creates a new TemplateCluster.
+// This cluster type supports config hot-reloading, but no self-healing of servers.
+// This cluster type does support hot-swapping of 2 online servers (which share the same index),
+// meaning that prior to swapping, the data for this vdisk will be copied from the old to the new server.
+// See `Cluster` for more information.
+func NewTemplateCluster(ctx context.Context, vdiskID string, optional bool, cs config.Source) (*Cluster, error) {
+	controller := &singleClusterStateController{
+		vdiskID:  vdiskID,
+		optional: optional,
+		// template cluster is read-only,
+		// hot-swapping is entirely an external responsibility
+		copyOnHotSwap: false,
+		configSource:  cs,
+		serverType:    log.ARDBTemplateServer,
+		getClusterID:  getTemplateClusterID,
+	}
+	err := controller.spawnConfigReloader(ctx, cs)
+	if err != nil {
+		controller.Close()
+		return nil, err
+	}
+
+	return NewCluster(vdiskID, controller)
+}
+
+type singleClusterStateController struct {
+	vdiskID   string
+	clusterID string
+
+	// when true, it means it's acceptable for the cluster not to exist
+	// otherwise this will be tracked as an error.
+	optional bool
+
+	// an optioan bool,
+	// when enabled data will be copied from the old to the new server,
+	// iff the new sever replaces the old server AND both are in state `online`
+	copyOnHotSwap bool
+
+	configSource config.Source
+
+	serverType log.ARDBServerType
 
 	servers     []config.StorageServerConfig
 	serverCount int64
 
-	pool   *ardb.Pool
+	mux sync.RWMutex
+
 	cancel context.CancelFunc
 
-	mux sync.RWMutex
+	getClusterID func(cfg config.VdiskNBDConfig) string
 }
 
-// Do implements StorageCluster.Do
-func (tsc *TemplateCluster) Do(_ ardb.StorageAction) (reply interface{}, err error) {
-	return nil, ErrMethodNotSupported
-}
+// ServerState implements ClusterStateController.ServerState
+func (ctrl *singleClusterStateController) ServerState() (state ServerState, err error) {
+	ctrl.mux.RLock()
+	defer ctrl.mux.RUnlock()
 
-// DoFor implements StorageCluster.DoFor
-func (tsc *TemplateCluster) DoFor(objectIndex int64, action ardb.StorageAction) (reply interface{}, err error) {
-	tsc.mux.RLock()
-	cfg, serverIndex, err := tsc.serverConfigFor(objectIndex)
-	tsc.mux.RUnlock()
+	if ctrl.serverCount == 0 {
+		err = ErrClusterNotDefined
+		return
+	}
+
+	state.Index, err = ardb.FindFirstServerIndex(ctrl.serverCount, ctrl.serverOperational)
 	if err != nil {
-		return nil, err
+		return
 	}
-	return tsc.doAt(serverIndex, cfg, action)
+
+	state.Config = ctrl.servers[state.Index]
+	state.Type = ctrl.serverType
+	return
 }
 
-// DoForAll implements StorageCluster.DoForAll
-func (tsc *TemplateCluster) DoForAll(pairs []ardb.IndexActionPair) ([]interface{}, error) {
-	// a shortcut in case we have received no pairs, or just a single one
-	switch len(pairs) {
-	case 0:
-		return nil, nil // nothing to do
-	case 1:
-		reply, err := tsc.DoFor(pairs[0].Index, pairs[1].Action)
-		if err != nil {
-			return nil, err
-		}
-		return []interface{}{reply}, nil
+// ServerStateFor implements ClusterStateController.ServerStateFor
+func (ctrl *singleClusterStateController) ServerStateFor(objectIndex int64) (state ServerState, err error) {
+	ctrl.mux.RLock()
+	defer ctrl.mux.RUnlock()
+
+	if ctrl.serverCount == 0 {
+		err = ErrClusterNotDefined
+		return
 	}
 
-	tsc.mux.RLock()
-	defer tsc.mux.RUnlock()
-
-	// sort all actions in terms of their mapped server index
-	servers := make(map[int64]*ardb.IndexActionMap)
-	for index, pair := range pairs {
-		// compute server index which maps to the given object index
-		serverIndex, err := ardb.ComputeServerIndex(tsc.serverCount, pair.Index, tsc.serverOperational)
-		if err != nil {
-			return nil, err
-		}
-		// add the pair to the relevant map
-		server, ok := servers[serverIndex]
-		if !ok {
-			server = new(ardb.IndexActionMap)
-			servers[serverIndex] = server
-		}
-		server.Add(int64(index), pair.Action)
+	state.Index, err = ardb.ComputeServerIndex(ctrl.serverCount, objectIndex, ctrl.serverOperational)
+	if err != nil {
+		return
 	}
 
-	// a shortcut if we are lucky enough to only have 1 server
-	if len(servers) == 1 {
-		for serverIndex, indexActionMap := range servers {
-			return ardb.Values(tsc.doAt(serverIndex, tsc.servers[serverIndex],
-				ardb.Commands(indexActionMap.Actions...)))
-		}
-	}
-
-	// create result -type and -channel to collect all replies async
-	type serverResult struct {
-		ServerIndex int64
-		Replies     []interface{}
-		Error       error
-	}
-	ch := make(chan serverResult)
-
-	// apply all actions async, and collect the results
-
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	var wg sync.WaitGroup
-	for serverIndex := range servers {
-		wg.Add(1)
-
-		// local variables to be used within the goroutine scope
-		indexActionMap := servers[serverIndex]
-		result := serverResult{ServerIndex: serverIndex}
-
-		go func() {
-			defer wg.Done()
-			result.Replies, result.Error = ardb.Values(tsc.doAt(
-				result.ServerIndex, tsc.servers[result.ServerIndex],
-				ardb.Commands(indexActionMap.Actions...)))
-			select {
-			case ch <- result:
-			case <-ctx.Done():
-			}
-		}()
-	}
-
-	// close channel when all inpu
-	go func() {
-		wg.Wait()
-		close(ch)
-	}()
-
-	// we'll collect all replies in order
-	replies := make([]interface{}, len(pairs))
-
-	// collect all replies
-	for result := range ch {
-		// return early if an error occured
-		if result.Error != nil {
-			return nil, errors.Wrapf(result.Error,
-				"error while applying actions in serverIndex %d", result.ServerIndex)
-		}
-
-		// get the indexActionMap for the given server,
-		// such that we can retrieve the correct reply index
-		m := servers[result.ServerIndex]
-
-		// collect all received replies in order
-		for i, reply := range result.Replies {
-			replyIndex := m.Indices[i]
-			replies[replyIndex] = reply
-		}
-	}
-
-	// return all replies from all servers in ordered form
-	return replies, nil
+	state.Config = ctrl.servers[state.Index]
+	state.Type = ctrl.serverType
+	return
 }
 
-// ServerIterator implements StorageCluster.ServerIterator.
-func (tsc *TemplateCluster) ServerIterator(context.Context) (<-chan ardb.StorageServer, error) {
-	return nil, ErrMethodNotSupported
+// ServerStateAt implements ClusterStateController.ServerStateAt
+func (ctrl *singleClusterStateController) ServerStateAt(serverIndex int64) (state ServerState, err error) {
+	ctrl.mux.RLock()
+	defer ctrl.mux.RUnlock()
+
+	if ctrl.serverCount == 0 {
+		err = ErrClusterNotDefined
+		return
+	}
+	if serverIndex < 0 || serverIndex >= ctrl.serverCount {
+		err = ardb.ErrServerIndexOOB
+		return
+	}
+
+	state.Index = serverIndex
+	state.Config = ctrl.servers[state.Index]
+	state.Type = ctrl.serverType
+	return
 }
 
-// ServerCount implements StorageCluster.ServerCount.
-func (tsc *TemplateCluster) ServerCount() int64 {
-	tsc.mux.RLock()
-
-	count := tsc.serverCount
-	for _, server := range tsc.servers {
-		if server.State != config.StorageServerStateOnline {
-			count--
-		}
+// UpdateServerState implements ClusterStateController.UpdateServerState
+func (ctrl *singleClusterStateController) UpdateServerState(state ServerState) bool {
+	ctrl.mux.Lock()
+	defer ctrl.mux.Unlock()
+	// ensure index is within range
+	if state.Index >= ctrl.serverCount {
+		log.Infof("couldn't update %s server for vdisk %s: index is OOB",
+			ctrl.serverType, ctrl.vdiskID)
+		return false // OOB
 	}
-	tsc.mux.RUnlock()
 
+	// ensure given config isn't out of date,
+	// as we'll assume that when the given (dial) config differs from the used config,
+	// the used config is correct and the given config is out of date.
+	if !storageServersEqual(ctrl.servers[state.Index], state.Config) {
+		log.Infof(
+			"couldn't update %s server #%d for vdisk %s: "+
+				"given config (%s) is out of date, it is now configured as %s",
+			ctrl.serverType, state.Index, ctrl.vdiskID, &state.Config,
+			&ctrl.servers[state.Index])
+		return false // given config != used config
+	}
+
+	// update server state (if the state is new)
+	return ctrl.setServerState(state.Index, state.Config.State)
+}
+
+// ServerCount implements ClusterStateController.ServerCount
+func (ctrl *singleClusterStateController) ServerCount() int64 {
+	ctrl.mux.RLock()
+	count := ctrl.serverCount
+	ctrl.mux.RUnlock()
 	return count
 }
 
-// Close any open resources
-func (tsc *TemplateCluster) Close() error {
-	tsc.cancel()
-	tsc.pool.Close()
+// Close implements ClusterStateController.Close
+func (ctrl *singleClusterStateController) Close() error {
+	ctrl.cancel()
 	return nil
 }
 
-func (tsc *TemplateCluster) doAt(serverIndex int64, cfg config.StorageServerConfig, action ardb.StorageAction) (reply interface{}, err error) {
-	conn, err := tsc.pool.Dial(cfg)
-	if err == nil {
-		defer conn.Close()
-		reply, err = action.Do(conn)
-		if err == nil || errors.Cause(err) == ardb.ErrNil {
-			return
-		}
+// serverOperational returns if a server is operational
+func (ctrl *singleClusterStateController) serverOperational(index int64) (bool, error) {
+	switch ctrl.servers[index].State {
+	case config.StorageServerStateOnline:
+		return true, nil
+	case config.StorageServerStateRIP:
+		return false, nil
+	default:
+		return false, ardb.ErrServerUnavailable
 	}
-
-	// an error has occured, broadcast it to AYS
-	status := MapErrorToBroadcastStatus(err)
-	log.Broadcast(
-		status,
-		log.SubjectStorage,
-		log.ARDBServerTimeoutBody{
-			Address:  cfg.Address,
-			Database: cfg.Database,
-			Type:     log.ARDBTemplateServer,
-			VdiskID:  tsc.vdiskID,
-		},
-	)
-
-	tsc.mux.Lock()
-	updateErr := tsc.updateServerState(serverIndex, config.StorageServerStateOffline)
-	tsc.mux.Unlock()
-	if updateErr != nil {
-		log.Errorf("couldn't update template server (%d) state to offline: %v", serverIndex, updateErr)
-	}
-
-	return nil, ardb.ErrServerUnavailable
 }
 
 // spawnConfigReloader starts all needed config watchers,
 // and spawns a goroutine to receive the updates.
 // An error is returned in case the initial watch-creation and config-update failed.
-// All future errors will be logged (and optionally broadcasted),
-// without stopping this goroutine.
-func (tsc *TemplateCluster) spawnConfigReloader(ctx context.Context, cs config.Source) error {
+// All future errors will be logged without stopping this goroutine.
+func (ctrl *singleClusterStateController) spawnConfigReloader(ctx context.Context, cs config.Source) error {
 	// create the context and cancelFunc used for the master watcher.
-	ctx, tsc.cancel = context.WithCancel(ctx)
+	ctx, ctrl.cancel = context.WithCancel(ctx)
 
 	// create the master watcher if possible
-	vdiskNBDRefCh, err := config.WatchVdiskNBDConfig(ctx, cs, tsc.vdiskID)
+	vdiskNBDRefCh, err := config.WatchVdiskNBDConfig(ctx, cs, ctrl.vdiskID)
 	if err != nil {
 		return err
 	}
 	vdiskNBDConfig := <-vdiskNBDRefCh
 
+	var clusterCfg config.StorageClusterConfig
+
 	// create the storage cluster watcher,
 	// and execute the initial config update iff
 	// an internal watcher is created.
-	var watcher ClusterConfigWatcher
-	clusterExists, err := watcher.SetClusterID(
-		ctx, cs, tsc.vdiskID, vdiskNBDConfig.TemplateStorageClusterID)
+	var clusterWatcher ClusterConfigWatcher
+	ctrl.clusterID = ctrl.getClusterID(vdiskNBDConfig)
+	clusterExists, err := clusterWatcher.SetClusterID(ctx, cs, ctrl.vdiskID, ctrl.clusterID)
 	if err != nil {
 		return err
 	}
-	var templateClusterCfg config.StorageClusterConfig
 	if clusterExists {
-		templateClusterCfg = <-watcher.Receive()
-		err = tsc.updateStorageConfig(templateClusterCfg)
-		if err != nil {
-			return err
-		}
+		clusterCfg = <-clusterWatcher.Receive()
+		ctrl.servers = clusterCfg.Servers
+		ctrl.serverCount = int64(len(clusterCfg.Servers))
+	} else if !ctrl.optional {
+		return errors.Wrapf(ErrClusterNotDefined,
+			"%s cluster %s does not exist", ctrl.serverType, ctrl.clusterID)
 	}
 
 	// spawn the config update goroutine
@@ -703,26 +497,24 @@ func (tsc *TemplateCluster) spawnConfigReloader(ctx context.Context, cs config.S
 					return
 				}
 
-				clusterWasDefined := watcher.Defined()
-				clusterExists, err = watcher.SetClusterID(
-					ctx, cs, tsc.vdiskID, vdiskNBDConfig.TemplateStorageClusterID)
+				ctrl.clusterID = ctrl.getClusterID(vdiskNBDConfig)
+				clusterExists, err := clusterWatcher.SetClusterID(ctx, cs, ctrl.vdiskID, ctrl.clusterID)
 				if err != nil {
-					log.Errorf("failed to watch new template cluster config: %v", err)
+					log.Errorf("failed to watch new %s cluster %s: %v", ctrl.serverType, ctrl.clusterID, err)
 					continue
 				}
-				if clusterWasDefined && !clusterExists {
-					// no cluster exists any longer, we need to delete the old state
-					tsc.mux.Lock()
-					tsc.servers, tsc.serverCount = nil, 0
-					tsc.mux.Unlock()
+				if !clusterExists {
+					if !ctrl.optional {
+						log.Errorf("%s cluster no longer exists, while it is required for vdisk %s",
+							ctrl.serverType, ctrl.vdiskID)
+						// [TODO] Notify AYS about this error
+					}
+					ctrl.setServers(nil)
 				}
 
 			// handle cluster storage updates
-			case templateClusterCfg = <-watcher.Receive():
-				err = tsc.updateStorageConfig(templateClusterCfg)
-				if err != nil {
-					log.Errorf("failed to update new template cluster config: %v", err)
-				}
+			case clusterCfg = <-clusterWatcher.Receive():
+				ctrl.setServers(clusterCfg.Servers)
 			}
 		}
 	}()
@@ -731,82 +523,117 @@ func (tsc *TemplateCluster) spawnConfigReloader(ctx context.Context, cs config.S
 	return nil
 }
 
-// updateStorageConfig overwrites the currently used storage config,
-// iff the given config is valid.
-func (tsc *TemplateCluster) updateStorageConfig(cfg config.StorageClusterConfig) error {
-	var clusterOperational bool
-	for _, server := range cfg.Servers {
-		if server.State == config.StorageServerStateOnline {
-			clusterOperational = true
-			break
+func (ctrl *singleClusterStateController) setServers(servers []config.StorageServerConfig) {
+	ctrl.mux.Lock()
+	defer ctrl.mux.Unlock()
+	serverCount := int64(len(servers))
+
+	// shrinking is now allowed, so handle this scenario
+	if serverCount < ctrl.serverCount {
+		// mark all "deleted" servers as state unknown, which will make this vdisk fail
+		// [TODO] Notify AYS about this error
+		for index := serverCount; index < ctrl.serverCount; index++ {
+			ctrl.servers[index].State = config.StorageServerStateUnknown
 		}
-	}
-	if !clusterOperational {
-		// no servers are available,
-		// so no need to use the config at all
-		tsc.mux.Lock()
-		tsc.servers, tsc.serverCount = nil, 0
-		tsc.mux.Unlock()
-		return nil
+	} else if serverCount > ctrl.serverCount { // growing isn't allowed either
+		// however if no serverCount was previously defined, we're OK with it
+		if ctrl.serverCount == 0 { // simply set servers without needing to do any handling
+			ctrl.servers = servers
+			ctrl.serverCount = serverCount
+			return
+		}
+
+		// ignore all extra servers
+		// [TODO] Notify AYS about this error
+		log.Errorf("hot-growing a cluster is not allowed, ignoring extra servers: %v", servers[ctrl.serverCount:])
+		// trim input servers so we don't get an overflow error
+		servers = servers[:ctrl.serverCount]
 	}
 
-	tsc.mux.Lock()
-	tsc.servers = cfg.Servers
-	tsc.serverCount = int64(len(cfg.Servers))
-	tsc.mux.Unlock()
-	return nil
+	// update all servers
+	var server, curServer config.StorageServerConfig
+	for index := range servers {
+		server, curServer = servers[index], ctrl.servers[index]
+
+		// check if dial-config changes
+		if !storageServersEqual(server, curServer) {
+			if curServer.State == config.StorageServerStateOnline && curServer.State == server.State {
+				// [TODO] Notify AYS about this error
+				log.Errorf(
+					"updating dial-config of online %s server #%d %s to %s, while hot-swapping is not supported (yet)",
+					ctrl.serverType, index, &curServer, &server)
+			} else {
+				// [TODO] Notify AYS about server dial-config update
+				log.Infof(
+					"updating dial-config of %s server #%d (state: %s) %s "+
+						"to %s (state: %s)",
+					ctrl.serverType, index, &curServer, curServer.State, &server, server.State)
+			}
+
+			ctrl.servers[index].Address = server.Address
+			ctrl.servers[index].Database = server.Database
+		}
+
+		// update state (if needed)
+		ctrl.setServerState(int64(index), server.State)
+	}
 }
 
-func (tsc *TemplateCluster) serverConfigFor(objectIndex int64) (cfg config.StorageServerConfig, serverIndex int64, err error) {
-	// ensure the template cluster is actually defined,
-	// as it is created even when no clusterID is referenced,
-	// just in case one would be defined via a hotreload.
-	if tsc.serverCount == 0 {
-		err = ErrClusterNotDefined
-		return
+// setServerState allows you to update the state of a pre-configured server,
+// the resulted boolean indicates whether or not an update took place.
+func (ctrl *singleClusterStateController) setServerState(index int64, state config.StorageServerState) bool {
+	old := ctrl.servers[index]
+	if old.State == state {
+		log.Debugf(
+			"ignoring update for %s server #%d %s (state: %s), as it remains unchanged",
+			ctrl.serverType, index, &old, old.State)
+		return false
 	}
 
-	// compute server index for the server which maps to the given object index
-	serverIndex, err = ardb.ComputeServerIndex(tsc.serverCount, objectIndex, tsc.serverOperational)
-	if err != nil {
-		return
+	if old.State == config.StorageServerStateRIP && state != config.StorageServerStateRIP {
+		// [TODO] Notify AYS about this error
+		log.Errorf("cannot change state of %s server #%d %s as it is marked as RIP, forcing state to RIP as well (instead of %s)",
+			ctrl.serverType, index, &old, state)
+		state = config.StorageServerStateRIP
 	}
 
-	// establish a connection for that serverIndex
-	cfg = tsc.servers[serverIndex]
-	return
-}
-
-func (tsc *TemplateCluster) updateServerState(index int64, state config.StorageServerState) error {
 	switch state {
-	case config.StorageServerStateOnline, config.StorageServerStateOffline, config.StorageServerStateRIP:
-
-		log.Debugf("updating vdisk %s' template server #%d state to %s", tsc.vdiskID, index, state)
-		tsc.servers[index].State = state
-		return nil
-
-	default:
-		return ardb.ErrServerStateNotSupported
-	}
-}
-
-// serverOperational returns true if
-// a server on the given index is available for operation.
-func (tsc *TemplateCluster) serverOperational(index int64) (bool, error) {
-	switch tsc.servers[index].State {
 	case config.StorageServerStateOnline:
-		return true, nil
+		// [TODO] Notify AYS about this (unexpected?) event
+		log.Infof(
+			"marking %s server #%d %s (state: %s) (back?) as online",
+			ctrl.serverType, index, &old, old.State)
 
 	case config.StorageServerStateOffline:
-		return false, ardb.ErrServerUnavailable
+		// [TODO] Notify AYS about this error
+		log.Errorf(
+			"marking %s server #%d %s (state: %s) as offline (no self-healing is possible)",
+			ctrl.serverType, index, &old, old.State)
 
 	case config.StorageServerStateRIP:
-		return false, nil
+		// [TODO] Notify AYS about this error
+		// [TODO] Support respreading of data (blocked by issue #601)
+		log.Errorf(
+			"marking %s server #%d %s (state: %s) as RIP (no respreading supported)",
+			ctrl.serverType, index, &old, old.State)
 
 	default:
-		return false, ardb.ErrServerStateNotSupported
+		// [TODO] Notify AYS about this error
+		log.Errorf(
+			"marking %s server #%d %s (state: %s) forcefully as Unknown, as state %s is not supported",
+			ctrl.serverType, index, &old, old.State, state)
+		state = config.StorageServerStateUnknown
 	}
+
+	ctrl.servers[index].State = state
+	return true
 }
+
+// getters to get a specific clusterID,
+// used to  create the different kind of singleCluster controllers.
+func getPrimaryClusterID(cfg config.VdiskNBDConfig) string  { return cfg.StorageClusterID }
+func getSlaveClusterID(cfg config.VdiskNBDConfig) string    { return cfg.SlaveStorageClusterID }
+func getTemplateClusterID(cfg config.VdiskNBDConfig) string { return cfg.TemplateStorageClusterID }
 
 // ClusterConfigWatcher is a small helper struct,
 // used to (un)set a storage cluster watcher for a given clusterID.
@@ -906,14 +733,19 @@ func MapErrorToBroadcastStatus(err error) log.MessageStatus {
 // enforces that our StorageClusters
 // are actually StorageClusters
 var (
-	_ ardb.StorageCluster = (*PrimaryCluster)(nil)
-	_ ardb.StorageCluster = (*TemplateCluster)(nil)
+	_ ardb.StorageCluster = (*Cluster)(nil)
 )
 
 // enforces that our ServerIterators
 // are actually ServerIterators
 var (
-	_ ardb.StorageServer = primaryStorageServer{}
+	_ ardb.StorageServer = smartServer{}
+)
+
+// enforces that our ClusterStateControllers
+// are actually ClusterStateControllers
+var (
+	_ ClusterStateController = (*singleClusterStateController)(nil)
 )
 
 var (
@@ -924,4 +756,17 @@ var (
 	// ErrClusterNotDefined is an error returned
 	// in case a cluster is used which is not defined.
 	ErrClusterNotDefined = errors.New("ARDB storage cluster is not defined")
+
+	// ErrServerNotDefined is returned when no server in a pair is defined
+	ErrServerNotDefined = errors.New("ARDB server is not defined")
+
+	// ErrServerIsDead is returned when an ARDB server (pair) is dead
+	ErrServerIsDead = errors.New("ARDB server (pair) is  dead")
+)
+
+// internal erros which should never propagate to the user
+var (
+	errActionNotApplied    = errors.New("storage action not applied")
+	errNopConfigUpdate     = errors.New("no config update was required")
+	errInvalidConfigUpdate = errors.New("invalid config update")
 )

--- a/nbd/ardb/storage/cluster_test.go
+++ b/nbd/ardb/storage/cluster_test.go
@@ -184,10 +184,13 @@ func TestPrimaryServerFailsByNotification(t *testing.T) {
 	sourceClusterConfig.Servers[1].State = config.StorageServerStateOffline
 	source.SetStorageCluster(clusterID, &sourceClusterConfig)
 	waitForAsyncClusterUpdate(t, func() bool {
-		cluster.mux.RLock()
-		ok := len(cluster.servers) == 2 && cluster.servers[1].State == config.StorageServerStateOffline
-		cluster.mux.RUnlock()
-		return ok
+		if cluster.controller.ServerCount() != 2 {
+			return false
+		}
+
+		state, err := cluster.controller.ServerStateAt(1)
+		require.NoError(err)
+		return state.Config.State == config.StorageServerStateOffline
 	})
 
 	// getting all content from the 1st server should still work
@@ -234,7 +237,7 @@ func TestTemplateServerFails(t *testing.T) {
 	require.NoError(err)
 	defer cluster.Close()
 
-	templateCluster, err := NewTemplateCluster(ctx, vdiskID, source)
+	templateCluster, err := NewTemplateCluster(ctx, vdiskID, false, source)
 	require.NoError(err)
 	defer templateCluster.Close()
 
@@ -278,10 +281,11 @@ func TestTemplateServerFails(t *testing.T) {
 	// create new primary servers
 	newSlice := redisstub.NewMemoryRedisSlice(2)
 	defer newSlice.Close()
-	cluster.mux.Lock()
-	cluster.servers = newSlice.StorageClusterConfig().Servers
-	cluster.serverCount = int64(len(cluster.servers))
-	cluster.mux.Unlock()
+	pcc := cluster.controller.(*singleClusterStateController)
+	pcc.mux.Lock()
+	pcc.servers = newSlice.StorageClusterConfig().Servers
+	pcc.serverCount = int64(len(pcc.servers))
+	pcc.mux.Unlock()
 
 	// getting all content from the 1st server should still work
 	for index := int64(0); index < blockCount; index += 2 {
@@ -327,7 +331,7 @@ func TestTemplateServerFailsByNotification(t *testing.T) {
 	require.NoError(err)
 	defer cluster.Close()
 
-	templateCluster, err := NewTemplateCluster(ctx, vdiskID, source)
+	templateCluster, err := NewTemplateCluster(ctx, vdiskID, false, source)
 	require.NoError(err)
 	defer templateCluster.Close()
 
@@ -369,19 +373,24 @@ func TestTemplateServerFailsByNotification(t *testing.T) {
 	templateClusterConfig.Servers[1].State = config.StorageServerStateOffline
 	source.SetStorageCluster(templateClusterID, &templateClusterConfig)
 	waitForAsyncClusterUpdate(t, func() bool {
-		templateCluster.mux.RLock()
-		ok := len(templateCluster.servers) == 2 && templateCluster.servers[1].State == config.StorageServerStateOffline
-		templateCluster.mux.RUnlock()
-		return ok
+		if templateCluster.controller.ServerCount() != 2 {
+			return false
+		}
+
+		state, err := templateCluster.controller.ServerStateAt(1)
+		require.NoError(err)
+		return state.Config.State == config.StorageServerStateOffline
 	})
 
 	// create new primary servers
 	newSlice := redisstub.NewMemoryRedisSlice(2)
 	defer newSlice.Close()
-	cluster.mux.Lock()
-	cluster.servers = newSlice.StorageClusterConfig().Servers
-	cluster.serverCount = int64(len(cluster.servers))
-	cluster.mux.Unlock()
+
+	pcc := cluster.controller.(*singleClusterStateController)
+	pcc.mux.Lock()
+	pcc.servers = newSlice.StorageClusterConfig().Servers
+	pcc.serverCount = int64(len(pcc.servers))
+	pcc.mux.Unlock()
 
 	// getting all content from the 1st server should still work
 	for index := int64(0); index < blockCount; index += 2 {
@@ -432,7 +441,7 @@ func TestTemplateServerRIP(t *testing.T) {
 	require.NoError(err)
 	defer cluster.Close()
 
-	templateCluster, err := NewTemplateCluster(ctx, vdiskID, source)
+	templateCluster, err := NewTemplateCluster(ctx, vdiskID, false, source)
 	require.NoError(err)
 	defer templateCluster.Close()
 
@@ -465,6 +474,10 @@ func TestTemplateServerRIP(t *testing.T) {
 	}
 }
 
+/*
+// [TODO]
+//    + Either activate these tests again, once this method is supported
+//    + or delete these tests and the underlying private test if this method is no longer required and defined
 func TestPrimaryClusterDoForAll(t *testing.T) {
 	slice := redisstub.NewMemoryRedisSlice(4)
 	defer slice.Close()
@@ -510,13 +523,14 @@ func TestTemplateClusterDoForAll(t *testing.T) {
 	cs.SetPrimaryStorageCluster(vdiskID, clusterID, nil)
 	cs.SetTemplateStorageCluster(vdiskID, clusterID, &cfg)
 
-	cluster, err := NewTemplateCluster(ctx, vdiskID, cs)
+	cluster, err := NewTemplateCluster(ctx, vdiskID, false, cs)
 	if err != nil {
 		t.Fatal(err)
 	}
 
 	testClusterDoForAll(t, cluster)
 }
+*/
 
 func testClusterDoForAll(t *testing.T, cluster ardb.StorageCluster) {
 	require := require.New(t)

--- a/nbd/nbdserver/backend_factory.go
+++ b/nbd/nbdserver/backend_factory.go
@@ -98,9 +98,9 @@ func (f *backendFactory) NewBackend(ctx context.Context, ec *nbd.ExportConfig) (
 
 	// create template cluster if supported by vdisk
 	// NOTE: internal template cluster may be nil, this is OK
-	var templateCluster *storage.TemplateCluster
+	var templateCluster *storage.Cluster
 	if staticConfig.Type.TemplateSupport() {
-		templateCluster, err = storage.NewTemplateCluster(ctx, vdiskID, f.configSource)
+		templateCluster, err = storage.NewTemplateCluster(ctx, vdiskID, true, f.configSource)
 		if err != nil {
 			resourceCloser.Close()
 			log.Error(err)

